eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -542,89 +549,187 @@ instance Semigroup ScriptResult where
 instance Monoid ScriptResult where
   mempty = Passes mempty
 
-data PlutusDebug
-  = PlutusDebugV1
-      CostModel
-      ExUnits
-      SBS.ShortByteString
-      [PV1.Data]
-      ProtVer
-  | PlutusDebugV2
-      CostModel
-      ExUnits
-      SBS.ShortByteString
-      [PV2.Data]
-      ProtVer
-  deriving (Eq, Generic, NoThunks)
+data PlutusData :: Language -> Type where
+  DataV1 :: {unDataV1 :: [PCD.Data]} -> PlutusData 'PlutusV1
+  DataV2 :: {unDataV2 :: [PCD.Data]} -> PlutusData 'PlutusV2
+
+deriving instance Eq (PlutusData l)
+
+mkPlutusData :: SLanguage l -> [PCD.Data] -> PlutusData l
+mkPlutusData = \case
+  SPlutusV1 -> DataV1
+  SPlutusV2 -> DataV2
+
+instance forall (l :: Language). (Typeable l, IsLanguage l) => ToCBOR (PlutusData l) where
+  toCBOR = \case
+    DataV1 dl -> toCBOR PlutusV1 <> encodeList toCBOR dl
+    DataV2 dl -> toCBOR PlutusV2 <> encodeList toCBOR dl
+
+instance forall (l :: Language). (Typeable l, IsLanguage l) => FromCBOR (PlutusData l) where
+  fromCBOR = do
+    _ <- fromCBOR @Language
+    mkPlutusData isLanguage <$> decodeList fromCBOR
+
+data PlutusDebug (l :: Language) where
+  PlutusDebug ::
+    { pdCostModel :: CostModel,
+      pdExUnits :: ExUnits,
+      pdSBS :: SBS.ShortByteString,
+      pdPlutusData :: PlutusData l,
+      pdProtVer :: ProtVer
+    } ->
+    PlutusDebug l
 
 -- | There is dummy Show instance for PlutusDebug intentionally, because it is too
 -- expensive and it will be too tempting to use it incorrectly. If needed for
 -- testing use 'StandaloneDeriving', otherwise define an efficient way to display
 -- this info.
-instance Show PlutusDebug where
+instance Show (PlutusDebug l) where
   show _ = "PlutusDebug Omitted"
 
-data PlutusError = PlutusErrorV1 PV1.EvaluationError | PlutusErrorV2 PV2.EvaluationError
+deriving instance Eq (PlutusDebug l)
+
+deriving instance Generic (PlutusDebug l)
+
+mkPlutusDebug ::
+  forall (l :: Language).
+  SLanguage l ->
+  CostModel ->
+  ExUnits ->
+  SBS.ShortByteString ->
+  [PCD.Data] ->
+  ProtVer ->
+  PlutusDebug l
+mkPlutusDebug sl costModel exUnits sbs pData protVer = case sl of
+  SPlutusV1 -> PlutusDebug costModel exUnits sbs (DataV1 pData) protVer
+  SPlutusV2 -> PlutusDebug costModel exUnits sbs (DataV2 pData) protVer
```

Comment:
This is no longer necessary, this will be exactly as the `PlutusDebug` constructor